danceradvent/public/articles/2020/23-asynchronous-promises.pod

```diff
@@ -0,0 +1,526 @@
+
+To get you started, I implemented the same exercise using L<IO::Async>.
+Luckily, because it already uses L<Future> (a more advanced version of
+the Promises pattern), we don't need to write one version with plain
+syntax and then another with Promises. It will provide us with this
+interface by default.
+
+    package CovidStats::IOAsync;
+
+    use Dancer2;
+    use experimental qw< postderef signatures >;
+
+    use DateTime;
+    use IO::Async;
+    use IO::Async::Loop;
+    use Future::Utils qw< fmap >;
+    use Net::Async::HTTP;
+    use Statistics::Descriptive::Full;
+    use URI::Escape qw< uri_escape >;
+
+    use constant {
+        'MAX_COUNTRIES' => 5,
+        'MAX_DAYS'      => 14,
+        'SUMMARY_URL'   => 'https://covid-api.mmediagroup.fr/v1/cases',
+        'COUNTRY_URL'   => 'https://covid-api.mmediagroup.fr/v1/history?country=%s&status=Confirmed',
+        'LTRIM'         => '0.25',
+    };
+
+    get '/' => sub {
+        return delayed {
+            flush();
+
+            my $loop = IO::Async::Loop->new();
+            my $http = Net::Async::HTTP->new();
+            $loop->add($http);
+
+            my $main_req = $http->GET( SUMMARY_URL() )->then( sub ($res) {
+                my $data      = from_json( $res->content );
+                my @countries = ( sort {
+                    $data->{$b}{'All'}{'confirmed'} <=> $data->{$a}{'All'}{'confirmed'}
+                } grep $_ ne 'Global', keys $data->%* )[ 0 .. MAX_COUNTRIES() ];
+
+                if ( !@countries ) {
+                    content("Sorry, failed to received countries\n");
+                    done();
+                    return;
+                }
+
+                return \@countries;
+            })->then( sub ($countries) {
+                my $yesterday = DateTime->now->subtract( 'days' => 1 );
+
+                fmap( sub ($country) {
+                    my $data_url = sprintf COUNTRY_URL(), uri_escape($country);
+
+                    return $http->GET($data_url)->then( sub ($res) {
+                        my $country_data = from_json( $res->content );
+
+                        my @period;
+                        my $day = $yesterday;
+                        foreach ( 0 .. MAX_DAYS() ) {
+                            push @period, $day->ymd();
+                            $day = $day->subtract( 'days' => 1 );
+                        }
+
+                        my @period_data = $country_data->{'All'}{'dates'}->@{@period};
+
+                        my $stat = Statistics::Descriptive::Full->new();
+                        $stat->add_data($_)
+                            for @period_data;
+
+                        return [ $country => $stat->trimmed_mean( LTRIM() ) ];
+                    })->catch( sub ($error) {
+                        content("Sorry, failed to fetch data for $country: $error");
+                    });
+                }, 'foreach'    => $countries,
+                   'concurrent' => scalar $countries->@*,
```

Comment:
A small thought: if this is a huge list, you'll do them all concurrently. The main reason to use `fmap` is to create a kind of concurrency that is bounded - not one at a time, but not all in a thundering herd. You might want to bound this to some limit, maybe 20?

```perl
use List::Util qw< max >;
...

    'concurrent' => max(20, scalar $countries->@*),
```

Reply:
I have a comment saying that this is not a proper practice in production, and that you should throttle yourself and that the API is very likely to do it for you.

I'm keeping this on "something you understand" rather than "production-level."